mlflow/models/evaluation/base.py

```diff
@@ -0,0 +1,559 @@
+            json.dump(self.metrics, fp)
+
+        artifacts_metadata = {
+            artifact_name: {
+                "uri": artifact.uri,
+                "class_name": _get_fully_qualified_class_name(artifact),
+            }
+            for artifact_name, artifact in self.artifacts.items()
+        }
+        with open(os.path.join(path, "artifacts_metadata.json"), "w") as fp:
+            json.dump(artifacts_metadata, fp)
+
+        artifacts_dir = os.path.join(path, "artifacts")
+        os.mkdir(artifacts_dir)
+
+        for artifact_name, artifact in self.artifacts.items():
+            artifact.save(os.path.join(artifacts_dir, artifact_name))
+
+    @property
+    def metrics(self) -> "mlflow.models.evaluation.EvaluationMetrics":
+        """
+        A dictionary mapping scalar metric names to scalar metric values
+        """
+        return self._metrics
+
+    @property
+    def artifacts(self) -> Dict[str, "mlflow.models.evaluation.EvaluationArtifact"]:
+        """
+        A dictionary mapping standardized artifact names (e.g. "roc_data") to
+        artifact content and location information
+        """
+        return self._artifacts
+
+
+_cached_mlflow_client = None
+
+
+class EvaluationDataset:
+    """
+    An input dataset for model evaluation. This is intended for use with the
+    :py:func:`mlflow.models.evaluation.evaluate()`
+    API.
+    """
+
+    NUM_SAMPLE_ROWS_FOR_HASH = 5
+    SPARK_DATAFRAME_LIMIT = 10000
+
+    def __init__(self, data, labels, name=None, path=None):
+        """
+        :param data: One of the following:
+         - A numpy array or list of evaluation features, excluding labels.
+         - A Pandas DataFrame, or a spark DataFrame,
+           containing evaluation features and labels. All columns will be regarded as feature
+           columns except the "labels" column.
+
+        :param labels: One of the following:
+         - A numpy array or list of evaluation labels, if `data` is also a numpy array or list.
+         - The string name of a column from `data` that contains evaluation labels, if `data`
+           is a DataFrame.
+
+        :param name: (Optional) The name of the dataset (must not contain ").
+
+        :param path: (Optional) the path to a serialized DataFrame (must not contain ").
+          (e.g. a delta table, parquet file)
+        """
+        import numpy as np
+        import pandas as pd
+
+        try:
+            from pyspark.sql import DataFrame as SparkDataFrame
+
+            supported_dataframe_types = (pd.DataFrame, SparkDataFrame)
+        except ImportError:
+            supported_dataframe_types = (pd.DataFrame,)
+
+        if name is not None and '"' in name:
+            raise ValueError(f'Dataset name cannot include a double quote (") but got {name}')
+        if path is not None and '"' in path:
+            raise ValueError(f'Dataset path cannot include a double quote (") but got {path}')
+
+        if isinstance(data, (np.ndarray, list)):
```

Comment:
should we allow sets here as well?

Reply:
according to design doc, should be only ndarray or list, one element in the array/list represent one example features data.